design/one-pager-resource-connectivity-mvp.md

```diff
@@ -638,7 +638,279 @@ parameters:
+* `AwsSecurityGroup`: allows the RDS instance to accept traffic from a certain IP and port.
+
+To achieve resource connectivity goal in AWS, two approaches can be imagined:
+
+1. **Configuration based single managed controller:** Encapsulate all the external resource types in a general `AwsNetwork` type and create a single managed resource for it.
+2. **External resource based managed controller**: Create a corresponding managed resource for each required external resource type
+
+Although the first approach requires potentially less effort, the latter approach provides a few advantages:
+
+1. **Reusability**: these managed resource types could be later re-used for other configuration
+1. **YAML Configuration**: instead of implementing a specific configuration logic inside `AwsNetwork`, the configuration details are expressed in a YAML file. This makes creating more sophisticated configurations possible, without having to write controllers. 
+
+This document focuses on implementing the second approach.  For doing so, we address the following challenges:
+
+* As pointed out [earlier](#resource-references) in this document, each managed resource can refer to other resources either by their external names, or their object reference. In AWS case, since we are taking a YAML configuration approach, interconnecting resource by api objects makes a lot of sense. At the same time it is still useful to re-use existing external resources. Because of these requirements, we introduce the notion of *reference type* as following:
+
+	```yaml
+	...
+	spec:
+	  ...
+	  someParameter:
+	    # one of {external, apiObject} values
+	    referenceType: apiObject
+      # a string of attributeName.namespacedName format (apiObject case), or arbitrary (external case)
+	    value: attributeNAme.sampleObject.sampleNamespace
+	  ...
+	...
+	```
+
+* One of the issues that rises in this approach, is the dependency that a managed resource might take on another managed resource, when the former uses the latter in its spec. For example, in order to create a `AwsSubnet` instance which consumes a `AwsVpc` instance, the `AwsVpc` instance has to be created and ready. To solve this, we introduce the notion of *resource dependency* in consuming resource's spec:
+	```yaml
+	...
+	metadata:
+	  - namespace: example
+	    name: sample
+	...
+	spec:
+	  ...
+    dependsOn:
+      # a list of strings, each in namespacedName format
+  		- requiredResourceName.requiredResourceNamespace
+	  ...
+	...
+	```
+
+This indicates that `sample.example` resource needs to wait until all required resources are ready. Moreoveer, if a resource refers to another resource as an `apiObject` explained above, its dependency is implicit.
+
+Putting all these together, the high fidelity of these managed resources can be modeled as following. 
+```yaml
+---
+apiVersion: network.aws.crossplane.io/v1alpha1
+kind: AwsVpc
```

Comment:
I think `Aws` can be omitted from all of these `kind` since they are part of the `network.aws.crossplane.io/v1alpha1` api.

Reply:
the reason I put `Aws` prefix there, is because it could be a bit confusing if you run `kubectl -n cross-plane get Vpc`, as it could be confused with `Vpc` type of other providers. Thoughts?

Reply:
I made the same comment separately; I agree that the `kubectl` behaviour is suboptimal, but I still prefer to omit the `Aws` prefix given these are namespaced within the `aws.crossplane.io` API. This is consistent with other resources with ambiguous names such as `Provider`.

Reply:
I agree that it is confusing, because if I recall correctly `kubectl` will only return a subset if there are multiple types. I don't have a great suggestion other than the two options which have already come up though.

Reply:
I don't think we will always have the ability or willingness to police stack resource names.  There is a user benefit to having consistency about _matching_ upstream resource names.  

I understand what you are trying to avoid because I have encountered the same problem with `kubectl get -A provider` and have needed to qualify the names.  That will likely become more common with multiple stacks installed where the resource names are common.

Reply:
I removed the `Aws` prefix.